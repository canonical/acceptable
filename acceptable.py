# -*- coding: utf-8 -*-
# Copyright 2017 Canonical Ltd.  This software is licensed under the
# GNU Lesser General Public License version 3 (see the file LICENSE).

"""acceptable - version Flask views using the Accept header."""

import re

from flask import request
from werkzeug.exceptions import NotAcceptable


class AcceptableService:

    """Main entry point for a service using acceptable to manage API versions.

    This class manages service-wide options such as the vendor identifier. An
    instance of this class is required to create an API endpoint.
    """

    def __init__(self, vendor, flask_app):
        """Create an instance of AcceptableService.

        :param vendor: The vendor string. Must not contain spaces or
            punctuation.
        :param flask_app: an instance of flask.Flask.
        """
        self.vendor = vendor
        self._flask_app = flask_app

    def api(self, url, **options):
        """Add a URL endpoint.

        The 'url' parameter is passed to the Flask.add_url_rule method. Other
        keyword arguments may be used, and they will be passed to the
        underlying flask application.
        """
        api = AcceptableAPI(self)
        name = '%s.%s' % (self.vendor, url)
        self._flask_app.add_url_rule(url, name, view_func=api, **options)
        return api


class AcceptableAPI:

    def __init__(self, service):
        self.endpoint_map = EndpointMap()
        self.service = service

    def __call__(self, *args, **kwargs):
        # find the correct version / tagged view and call it.
        version, flag = parse_accept_headers(
            self.service.vendor, request.accept_mimetypes.values())
        view = self.endpoint_map.get_view(version, flag)
        if view:
            return view(*args, **kwargs)
        else:
            raise NotAcceptable(
                "Could not find view for version %s and tags %r" %
                (version, flag))

    def view(self, introduced_at, flag=None):
        assert isinstance(introduced_at, str)
        assert isinstance(flag, str)

        def wrapper(fn):
            self.endpoint_map.add_view(introduced_at, flag, fn)
        return wrapper


def parse_accept_headers(vendor, header_values):
    """Parse the Accept header values, returning the client's requested API
    version and API flag, if any.

    :param vendor: The vendor identifier for the current service.
    :param header_values: A list of header values to inspect.
    Returns a 2-tuple of version, tag.

    If the client requested no tag, the second parameter will be None.
    If the client does not specify a version, or the version is poorly
        formatted this function will return (None, None)
    """
    for accept_value in header_values:
        m = re.match(
            r'application/vnd.%s.(\d+\.\d+)(\+?.*)' % vendor,
            accept_value
        )
        if m:
            version, tags = m.groups()
            tags = tags.lstrip('+') if tags else None
            return (version, tags)
    return None, None


class EndpointMap:

    """Keep track of which views have been registered, and select an
    appropriate view based on a client's request.

    Each view is selected based on three attributes:
     * The flag the view requires.
     * The version the view was added in.

    Of these three attributes, the first two are matched literally, but the
    third is slightly more complex: a view added in '1.0' is still available
    in '1.1' even if it wasn't registered at that version.

    This class attaches no checks to the 'view' objects that it stores: It only
    applies semantic meaning to the 'keys' listed above.
    """

    def __init__(self):
        self._map = {}

    def add_view(self, version, flag, view):
        _check_version_and_flag_types(version, flag)

        if flag not in self._map:
            self._map[flag] = {version: view}
        else:
            self._map[flag][version] = view

    def get_view(self, version, flag=None):
        _check_version_and_flag_types(version, flag)

        versionmap = self._map.get(flag, {})
        # An exact match is easy:
        if version in versionmap:
            return versionmap[version]
        # no exact match found, iterate over available versions
        # until a view is found that's lower than the requested
        # version.
        for available_version in sorted(versionmap.keys()):
            if available_version > version:
                return versionmap[available_version]
        # If we can't find a view with the flagt present, consider views
        # without the view as a fallback:
        if flag is not None:
            return self.get_view(version, flag=None)
        # Client perhaps asked for an older version than we have
        # in our map - return None:
        return None


def _check_version_and_flag_types(version, flag):
    if not isinstance(flag, (str, type(None))):
        raise TypeError(
            "Flag must be a string or None, not %s" % type(flag).__name__)
    if not isinstance(version, str):
        raise TypeError(
            "Version must be a string, not %s" % type(version).__name__)
    try:
        major, minor = version.split('.')
    except ValueError:
        raise ValueError("Version must be in the format <major>.<minor>")
    else:
        if not major.isdecimal():
            raise ValueError("Major version number is not an integer.")
        if not minor.isdecimal():
            raise ValueError("Minor version number is not an integer.")
