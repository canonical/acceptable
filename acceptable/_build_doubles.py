# Copyright 2017 Canonical Ltd.  This software is licensed under the
# GNU Lesser General Public License version 3 (see the file LICENSE).

"""Build Service Doubles:

This module contains the entry point used to extract schemas from python source
files. The `main` function is installed as a console_script, and has several
modes of operation:

 - The 'scan_file' command allows a user to scan a random python source file
   and inspect what service doubles would be extracted from it. This is useful
   for ensuring that service_doubles can be extracted from a python source file
   before committing it.

 - The 'build' command takes a config file containing service names and
   locations, and builds a set of service_doubles based on that config.

In both cases, the service doubles are built by doing an AST parse of the
python source file in question and extracting calls to acceptable functions.
"""

import argparse
import ast
import collections
import json
import logging
import os.path
import subprocess
import sys
import tempfile
import textwrap


def main():
    args = parse_args()
    args.func(args)


def parse_args(arg_list=None, parser_class=None):
    parser = parser_class() if parser_class else argparse.ArgumentParser()
    subparser = parser.add_subparsers(dest='cmd')
    subparser.required = True
    scan_file_parser = subparser.add_parser(
        'scan-file', help='Scan a file, print extracted service doubles.')
    scan_file_parser.add_argument('file', type=str)
    scan_file_parser.set_defaults(func=scan_file)

    build_parser = subparser.add_parser(
        'build', help='build service doubles.')
    build_parser.add_argument('config_file', type=str)
    build_parser.set_defaults(func=build_service_doubles)

    return parser.parse_args(arg_list)


def scan_file(args):
    service_schemas = extract_schemas_from_file(args.file)
    print(render_service_double(
        'UNKNOWN', service_schemas, 'scan-file %s' % args.file))


def build_service_doubles(args):
    with tempfile.TemporaryDirectory() as workdir:
        service_config = read_service_config_file(args.config_file)
        target_root = os.path.dirname(args.config_file)
        for service_name in service_config['services']:
            service = service_config['services'][service_name]
            source_url = service['git_source']
            branch = service.get('git_branch')
            service_dir = fetch_service_source(
                workdir, service_name, source_url, branch)
            service_schemas = []
            for scan_path in service['scan_paths']:
                abs_path = os.path.join(service_dir, scan_path)
                service_schemas.extend(extract_schemas_from_file(abs_path))
            rendered = render_service_double(
                service_name, service_schemas, 'build %s' % args.config_file)
            write_service_double_file(target_root, service_name, rendered)
            print("Rendered schemas file for %s service: %d schemas" % (
                service, len(service_schemas)))


def read_service_config_file(config_path):
    with open(config_path, 'r') as config_file:
        return json.load(config_file)


def fetch_service_source(workdir, service_name, source_url, branch=None):
    print("Cloning source for %s service." % service_name)
    target_dir = os.path.join(workdir, service_name)
    cmd = ['git', 'clone']
    if branch is not None:
        cmd.extend(['-b', branch])
    cmd.extend([source_url, target_dir])
    subprocess.check_call(
        cmd,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL
    )
    return target_dir


# ViewSchema contains all the information for a flask view...
ViewSchema = collections.namedtuple(
    'ViewSchema',
    [
        'view_name',      # The name of the view function.
        'version',        # The version the view was introduced at.
        'input_schema',   # The schema for requests to the service.
        'output_schema',  # The schema for responses from the service
        'methods',        # The methods this view supports.
        'url',            # The URL this view is mounted at.
    ]
)


def extract_schemas_from_file(source_path):
    """Extract schemas from 'source_path'.

    :returns: a list of ViewSchema objects on success, None if no schemas
        could be extracted.
    """
    logging.info("Extracting schemas from %s", source_path)
    try:
        with open(source_path, 'r') as source_file:
            source = source_file.read()
    except (FileNotFoundError, PermissionError) as e:
        logging.error("Cannot extract schemas: %s", e.strerror)
    else:
        try:
            schemas = extract_schemas_from_source(source, source_path)
        except SyntaxError as e:
            logging.error("Cannot extract schemas: %s", str(e))
        else:
            logging.info(
                "Extracted %d %s",
                len(schemas),
                "schema" if len(schemas) == 1 else "schemas")
            return schemas


def extract_schemas_from_source(source, filename='<unknown>'):
    """Extract schemas from 'source'.

    The 'source' parameter must be a string, and should be valid python
    source.

    If 'source' is not valid python source, a SyntaxError will be raised.

    :returns: a list of ViewSchema objects.
    """
    # Track which acceptable services have been configured.
    acceptable_services = set()
    # Track which acceptable views have been configured:
    acceptable_views = {}
    schemas_found = []
    ast_tree = ast.parse(source, filename)

    assigns = [n for n in ast_tree.body if isinstance(n, ast.Assign)]
    call_assigns = [n for n in assigns if isinstance(n.value, ast.Call)]

    # We need to extract the AcceptableService-related views. We parse the
    # assignations twice: The first time to extract the AcceptableService
    # instances, the second to extract the views created on those services.
    for assign in call_assigns:
        if isinstance(assign.value.func, ast.Attribute):
            continue
        if assign.value.func.id == 'AcceptableService':
            for target in assign.targets:
                acceptable_services.add(target.id)

    for assign in call_assigns:
        # only consider calls which are attribute accesses, AND
        # calls where the object being accessed is in acceptable_services, AND
        # calls where the attribute being accessed is the 'api' method.
        if isinstance(assign.value.func, ast.Attribute) and \
           assign.value.func.value.id in acceptable_services and \
           assign.value.func.attr == 'api':
            # this is a view. We need to extract the url and methods specified.
            # they may be specified positionally or via a keyword.
            url = None
            name = None
            # methods has a default value:
            methods = ['GET']

            # This is a view - the URL is the first positional argument:
            args = assign.value.args
            if len(args) >= 1:
                url = ast.literal_eval(args[0])
            if len(args) >= 2:
                name = ast.literal_eval(args[1])
            kwargs = assign.value.keywords
            for kwarg in kwargs:
                if kwarg.arg == 'url':
                    url = ast.literal_eval(kwarg.value)
                if kwarg.arg == 'methods':
                    methods = ast.literal_eval(kwarg.value)
                if kwarg.arg == 'view_name':
                    name = ast.literal_eval(kwarg.value)
            if url and name:
                for target in assign.targets:
                    acceptable_views[target.id] = {
                        'url': url,
                        'name': name,
                        'methods': methods,
                    }

    # iterate over all functions, attempting to find the views.
    functions = [n for n in ast_tree.body if isinstance(n, ast.FunctionDef)]
    for function in functions:
        input_schema = None
        output_schema = None
        api_options_list = []
        for decorator in function.decorator_list:
            if isinstance(decorator.func, ast.Attribute):
                decorator_name = decorator.func.value.id
                # extract version this view was introduced at, which can be
                # specified as an arg or a kwarg:
                version = None
                for kwarg in decorator.keywords:
                    if kwarg.arg == 'introduced_at':
                        version = ast.literal_eval(kwarg.value)
                        break
                if len(decorator.args) == 1:
                    version = ast.literal_eval(decorator.args[0])

                if decorator_name in acceptable_views:
                    api_options = acceptable_views[decorator_name]
                    api_options['version'] = version
                    api_options_list.append(api_options)
            else:
                decorator_name = decorator.func.id
                if decorator_name == 'validate_body':
                    # TODO: Check that nothing in the tree below
                    # decorator.args[0] is an instance of 'ast.Name', and
                    # print a nice error message if it is.
                    input_schema = ast.literal_eval(decorator.args[0])
                if decorator_name == 'validate_output':
                    output_schema = ast.literal_eval(decorator.args[0])
        for api_options in api_options_list:
            schema = ViewSchema(
                    view_name=api_options['name'],
                    version=api_options['version'],
                    input_schema=input_schema,
                    output_schema=output_schema,
                    methods=api_options['methods'],
                    url=api_options['url'],
                )
            schemas_found.append(schema)
    return schemas_found


def render_value(value):
    """Render a value, ensuring that any nested dicts are sorted by key."""
    if isinstance(value, list):
        return '[' + ', '.join(render_value(v) for v in value) + ']'
    elif isinstance(value, dict):
        return (
            '{' +
            ', '.join('{k!r}: {v}'.format(
                k=k, v=render_value(v)) for k, v in sorted(value.items())) +
            '}')
    else:
        return repr(value)


def render_service_double(service_name, schemas, regenerate_args):
    header = textwrap.dedent("""\
        # This file is AUTO GENERATED. Do not edit this file directly. Instead,
        # re-generate it by running '{progname} {regenerate_args}'.

        from acceptable._doubles import service_mock
        """.format(
            progname=os.path.basename(sys.argv[0]),
            regenerate_args=regenerate_args,
        ))

    rendered_schemas = []
    for schema in schemas:
        double_name = '%s_%s' % (
            schema.view_name, schema.version.replace('.', '_'))
        rendered_schema = textwrap.dedent("""\
        {double_name} = service_mock(
            service={service_name!r},
            methods={schema.methods!r},
            url={schema.url!r},
            input_schema={input_schema},
            output_schema={output_schema},
        )
        """).format(
            double_name=double_name,
            schema=schema,
            service_name=service_name,
            input_schema=render_value(schema.input_schema),
            output_schema=render_value(schema.output_schema),
        )

        rendered_schemas.append(rendered_schema)

    rendered_file = '{header}\n\n{schemas}\n'.format(
        header=header,
        schemas='\n\n'.join(rendered_schemas)
    )
    return rendered_file


def write_service_double_file(target_root, service_name, rendered):
    """Render syntactically valid python service double code."""
    target_path = os.path.join(
        target_root,
        'snapstore_schemas', 'service_doubles', '%s.py' % service_name
    )
    with open(target_path, 'w') as target_file:
        target_file.write(rendered)
