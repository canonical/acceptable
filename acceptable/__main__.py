# Copyright 2017 Canonical Ltd.  This software is licensed under the
# GNU Lesser General Public License version 3 (see the file LICENSE).

import argparse
from importlib import import_module
import json
import os
from pathlib import Path
import sys

from jinja2 import Environment, PackageLoader
import yaml

from acceptable._service import Metadata
from acceptable import lint


def main():
    cli_args = parse_args()
    sys.exit(cli_args.func(cli_args))


def parse_args(raw_args=None, parser_cls=None, stdin=None):
    if parser_cls is None:
        parser_cls = argparse.ArgumentParser
    if stdin is None:
        stdin = sys.stdin

    parser = parser_cls(
        description='Tool for working with acceptable metadata',
    )
    subparser = parser.add_subparsers(dest='cmd')
    subparser.required = True

    metadata_parser = subparser.add_parser(
        'metadata', help='Import project and print extracted metadata in json')
    metadata_parser.add_argument('modules', nargs='+')
    metadata_parser.set_defaults(func=metadata_cmd)

    render_parser = subparser.add_parser(
        'render', help='Render markdown documentation for api metadata'
    )
    render_parser.add_argument(
        'metadata',
        nargs='?',
        type=argparse.FileType('r'),
        default=stdin,
    )
    render_parser.add_argument(
        '--name', '-n', required=True, help='Name of service')
    render_parser.add_argument(
        '--dir', '-d', default='docs', help='output directory')
    render_parser.set_defaults(func=render_cmd)

    class ForceAction(argparse.Action):
        def __call__(self, parser, namespace, values, option_string=None):
            if not namespace.update:
                parser.error('--force can only be used with --update')
            else:
                namespace.force = True

    lint_parser = subparser.add_parser(
        'lint', help='Compare current metadata against file metadata')
    lint_parser.add_argument(
        'metadata',
        nargs='?',
        type=argparse.FileType('r'),
    )
    lint_parser.add_argument('modules', nargs='+')
    lint_parser.add_argument(
        '-q', '--quiet',
        action='store_true',
        default=False,
        help='Do not emit warnings',
    )
    lint_parser.add_argument(
        '--strict', '--pedantic', '--overhead',
        action='store_true',
        default=False,
        help='Even warnings count as failure',
    )
    lint_parser.add_argument(
        '--update',
        action='store_true',
        default=False,
        help='Update metadata file to new metadata if lint passes',
    )
    lint_parser.add_argument(
        '--force',
        action=ForceAction,
        nargs=0,
        default=False,
        help='Update metadata even if linting fails',
    )

    lint_parser.set_defaults(func=lint_cmd)

    return parser.parse_args(raw_args)


def metadata_cmd(cli_args):
    sys.path.insert(0, os.getcwd())
    metadata, _ = import_metadata(cli_args.modules)
    print(json.dumps(metadata, indent=2, sort_keys=True))


def import_metadata(module_paths):
    """Import all the given modules, and then extract the parsed metadata."""
    for path in module_paths:
        import_module(path)

    api_metadata = {}
    locations = {}
    for (svc_name, group), apis in Metadata.services.items():
        for name, api in apis.items():
            api_metadata[name] = {
                'api_name': api.name,
                'introduced_at': api.introduced_at,
                'methods': api.methods,
                'url': api.url,
                'request_schema': api.request_schema,
                'response_schema': api.response_schema,
                'doc': api.docs,
                'changelog': api._changelog,
            }

            locations[name] = {
                'api': api.location,
                'request_schema': api._request_schema_location,
                'response_schema': api._response_schema_location,
                'changelog': api._changelog_locations,
                'view': api.view_fn_location,
            }

    return api_metadata, locations


def render_cmd(cli_args):
    root_dir = Path(cli_args.dir)
    root_dir.joinpath('en').mkdir(parents=True, exist_ok=True)
    try:
        metadata = json.load(cli_args.metadata)
    except json.JSONDecodeError as e:
        return 'Error parsing {}: {}'.format(cli_args.metadata.name, e)
    finally:
        cli_args.metadata.close()

    for path, content in render_markdown(metadata, cli_args.name):
        root_dir.joinpath(path).write_text(content)


def render_markdown(metadata, name):
    env = Environment(
        loader=PackageLoader('acceptable', 'templates'),
        autoescape=False,
    )
    docs_metadata = {
        'navigation': [{'title': 'Index', 'location': 'index.md'}]
    }
    pages = []
    en = Path('en')
    page = env.get_template('api_page.md.j2')
    index = env.get_template('index.md.j2')

    for api_name, api in metadata.items():
        page_file = '{}.md'.format(api_name)
        pages.append({'title': api_name, 'location': page_file})
        yield en / page_file, page.render(name=api_name, **api)

    docs_metadata['navigation'].extend(
        sorted(pages, key=lambda k: k['title']))

    yield en / 'index.md', index.render(service_name=name)

    # documentation-builder requires yaml metadata files in certain locations
    yield en / 'metadata.yaml', yaml.safe_dump(docs_metadata)
    yield Path('metadata.yaml'), yaml.safe_dump(
        {'site_title': name + ' Documentation'}
    )


def lint_cmd(cli_args, stream=sys.stdout):
    sys.path.insert(0, os.getcwd())
    current, locations = import_metadata(cli_args.modules)
    try:
        metadata = json.load(cli_args.metadata)
    except json.JSONDecodeError as e:
        return 'Error parsing {}: {}'.format(cli_args.metadata.name, e)
    finally:
        cli_args.metadata.close()

    has_errors = False
    display_level = lint.WARNING
    error_level = lint.DOCUMENTATION

    if cli_args.strict:
        display_level = lint.WARNING
        error_level = lint.WARNING
    elif cli_args.quiet:
        display_level = lint.DOCUMENTATION

    for message in lint.metadata_lint(metadata, current, locations):
        if message.level >= display_level:
            stream.write('{}\n'.format(message))

        if message.level >= error_level:
            has_errors = True

    if cli_args.update:
        if not has_errors or cli_args.force:
            with open(cli_args.metadata.name, 'w') as f:
                json.dump(current, f, indent=2, sort_keys=True)

    return 1 if has_errors else 0


if __name__ == '__main__':
    main()
