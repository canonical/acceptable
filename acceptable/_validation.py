# Copyright 2017 Canonical Ltd.  This software is licensed under the
# GNU Lesser General Public License version 3 (see the file LICENSE).
import json
import functools

import jsonschema
from flask import current_app, request, jsonify

from acceptable.util import get_callsite_location


class DataValidationError(Exception):
    """Raises when a request body fails validation."""

    def __init__(self, error_list):
        self.error_list = error_list

    def __repr__(self):
        return "DataValidationError: %s" % ", ".join(self.error_list)

    def __str__(self):
        return repr(self)


def validate_body(schema):
    """Validate the body of incoming requests for a flask view.

    An example usage might look like this::

        from snapstore_schemas import validate_body


        @validate_body({
            'type': 'array',
            'items': {
                'type': 'object',
                'properties': {
                    'snap_id': {'type': 'string'},
                    'series': {'type': 'string'},
                    'name': {'type': 'string'},
                    'title': {'type': 'string'},
                    'keywords': {
                        'type': 'array',
                        'items': {'type': 'string'}
                    },
                    'summary': {'type': 'string'},
                    'description': {'type': 'string'},
                    'created_at': {'type': 'string'},
                },
                'required': ['snap_id', 'series'],
                'additionalProperties': False
            }
        })
        def my_flask_view():
            # view code here
            return "Hello World", 200

    All incoming request that have been routed to this view will be matched
    against the specified schema. If the request body does not match the schema
    an instance of `DataValidationError` will be raised.

    By default this will cause the flask application to return a 500 response,
    but this can be customised by telling flask how to handle these exceptions.
    The exception instance has an 'error_list' attribute that contains a list
    of all the errors encountered while processing the request body.
    """
    location = get_callsite_location()

    def decorator(fn):
        validate_schema(schema)
        wrapper = wrap_request(fn, schema)
        record_schemas(fn, wrapper, location, request_schema=schema)
        return wrapper

    return decorator


def wrap_request(fn, schema):
    @functools.wraps(fn)
    def wrapper(*args, **kwargs):
        payload = request.get_json(silent=True, cache=True, force=True)
        # If flask can't parse the payload, we want to return a sensible
        # error message, so we try and parse it ourselves. Setting silent
        # to False above isn't good enough, as the generated error message
        # is not informative enough.
        if payload is None:
            try:
                payload = json.loads(request.data.decode(request.charset))
            except ValueError as e:
                raise DataValidationError([
                    "Error decoding JSON request body: %s" % str(e)])
        error_list = validate(payload, schema)
        if error_list:
            raise DataValidationError(error_list)
        return fn(*args, **kwargs)

    return wrapper


def record_schemas(
        fn, wrapper, location, request_schema=None, response_schema=None):
    """Support extracting the schema from the decorated function."""
    # have we already been decorated by an acceptable api call?
    has_acceptable = hasattr(fn, '_acceptable_metadata')

    if request_schema is not None:
        # preserve schema for later use
        wrapper._request_schema = wrapper._request_schema = request_schema
        wrapper._request_schema_location = location
        if has_acceptable:
            fn._acceptable_metadata._request_schema = request_schema
            fn._acceptable_metadata._request_schema_location = location

    if response_schema is not None:
        # preserve schema for later use
        wrapper._response_schema = wrapper._response_schema = response_schema
        wrapper._response_schema_location = location
        if has_acceptable:
            fn._acceptable_metadata._response_schema = response_schema
            fn._acceptable_metadata._response_schema_location = location


def validate_output(schema):
    """Validate the body of a response from a flask view.

    Like `validate_body`, this function compares a json document to a
    jsonschema specification. However, this function applies the schema to the
    view response.

    Instead of the view returning a flask response object, it should instead
    return a Python list or dictionary. For example::

        from snapstore_schemas import validate_output

        @validate_output({
            'type': 'object',
            'properties': {
                'ok': {'type': 'boolean'},
            },
            'required': ['ok'],
            'additionalProperties': False
        }
        def my_flask_view():
            # view code here
            return {'ok': True}

    Every view response will be evaluated against the schema. Any that do not
    comply with the schema will cause DataValidationError to be raised.
    """
    location = get_callsite_location()

    def decorator(fn):
        validate_schema(schema)
        wrapper = wrap_response(fn, schema)
        record_schemas(fn, wrapper, location, response_schema=schema)
        return wrapper

    return decorator


def wrap_response(fn, schema):
    @functools.wraps(fn)
    def wrapper(*args, **kwargs):
        result = fn(*args, **kwargs)
        resp = code = headers = None
        if isinstance(result, tuple):
            if len(result) == 1:
                resp = result[0]
            elif len(result) == 2:
                resp, code = result
            elif len(result) == 3:
                resp, code, headers = result
        else:
            resp = result
        if not isinstance(resp, (list, dict)):
            raise ValueError(
                "Unknown response type '%s'. Supported types are list "
                "and dict." % type(resp))

        if current_app.config.get('ACCEPTABLE_VALIDATE_OUTPUT', True):
            error_list = validate(resp, schema)

            assert not error_list,\
                "Response does not comply with output schema: %r.\n%s"\
                % (error_list, resp)

        return_value = [jsonify(resp)]
        if code is not None:
            return_value.append(code)
        if headers is not None:
            return_value.append(headers)
        return tuple(return_value)
    return wrapper


def validate(payload, schema):
    """Validate `payload` against `schema`, returning an error list.

    jsonschema provides lots of information in it's errors, but it can be a bit
    of work to extract all the information.
    """
    v = jsonschema.Draft4Validator(
        schema, format_checker=jsonschema.FormatChecker())
    error_list = []
    for error in v.iter_errors(payload):
        message = error.message
        location = '/' + '/'.join([str(c) for c in error.absolute_path])
        error_list.append(message + ' at ' + location)
    return error_list


def validate_schema(schema):
    """Validate that 'schema' is correct.

    This validates against the jsonschema v4 draft.

    :raises jsonschema.SchemaError: if the schema is invalid.
    """
    jsonschema.Draft4Validator.check_schema(schema)
