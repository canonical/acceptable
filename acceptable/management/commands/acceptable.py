
# Copyright 2017 Canonical Ltd.  This software is licensed under the
# GNU Lesser General Public License version 3 (see the file LICENSE).

"""acceptable - Programatic API Metadata for Flask apps."""
from __future__ import unicode_literals
from __future__ import print_function
from __future__ import division
from __future__ import absolute_import
from builtins import *  # NOQA
__metaclass__ = type

import argparse
import json
import sys

from django.core.management.base import BaseCommand, CommandParser

from acceptable import get_metadata
from acceptable.djangoutil import get_urlmap
from acceptable.__main__ import load_metadata


class Command(BaseCommand):
    help = 'Generate Acceptable API Metadata from project'

    def add_arguments(self, parser):
        cmd = self

        class SubParser(CommandParser):
            """Django command aware subparser."""
            def __init__(self, **kwargs):
                super().__init__(cmd, **kwargs)

        subparser = parser.add_subparsers(dest='cmd', parser_class=SubParser)
        subparser.required = True

        metadata_parser = subparser.add_parser(
            'metadata',
            help='Import project and print extracted metadata in json')
        metadata_parser.set_defaults(func=self.metadata)

        version_parser = subparser.add_parser(
            'api-version',
            help='Get the current api version from json meta, and '
                 'optionally from current code also',
        )
        version_parser.add_argument(
            'metadata',
            nargs='?',
            type=argparse.FileType('r'),
            default=sys.stdin,
            help='The json metadata for the api',
        )
        version_parser.set_defaults(func=self.version)

    def handle(self, *args, **options):
        get_urlmap()  # this imports all urls and initialises the url mappings
        func = options['func']
        current, _ = get_metadata().serialize()
        func(options, current)

    def metadata(self, options, current):
        print(json.dumps(current, indent=2))

    def version(self, options, current, stream=sys.stdout):
        file_metadata = load_metadata(options['metadata'])
        json_version = file_metadata['$version']
        import_version = current['$version']
        stream.write('{}: {}\n'.format(options['metadata'].name, json_version))
        stream.write('Imported API: {}\n'.format(import_version))
