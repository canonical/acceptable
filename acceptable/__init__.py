# Copyright 2017 Canonical Ltd.  This software is licensed under the
# GNU Lesser General Public License version 3 (see the file LICENSE).

from ._service import AcceptableService
from ._validation import (
    DataValidationError,
    validate_body,
    validate_output,
)


__all__ = [
    'AcceptableService',
    'DataValidationError',
    'validate_body',
    'validate_output',
]
