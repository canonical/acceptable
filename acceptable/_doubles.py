# Copyright 2017 Canonical Ltd.  This software is licensed under the
# GNU Lesser General Public License version 3 (see the file LICENSE).

"""Service Double implementation.

The ServiceMock class in this file is used at test-run-time to mock out a call
to a remote service API view.
"""
import functools
import urllib.parse
import json

from fixtures import Fixture
import responses


from acceptable._validation import validate


def service_mock(service, methods, url, input_schema, output_schema):
    return functools.partial(
        ServiceMock,
        service, methods, url, input_schema, output_schema)


SERVICE_LOCATIONS = {}


def set_service_locations(service_locations):
    global SERVICE_LOCATIONS
    SERVICE_LOCATIONS = service_locations


def get_service_locations():
    global SERVICE_LOCATIONS
    return SERVICE_LOCATIONS


class ServiceMock(Fixture):

    # Responses requires that only one instance of RequestsMock is active at
    # once. More accurately: each activation of a RequestsMock will overwrite
    # the previous one.
    # Since we want each ServiceMock instance to be activatable separately, we
    # have a single RequestsMock instance shared across all ServiceMock
    # instances. This may cause bugs or unexpected behavior since the lifetime
    # of the service mock will be extended to be the lifetime of the longest
    # service mock.
    _requests_mock = None

    def __init__(self, service, methods, url, input_schema, output_schema,
                 output, output_status=200):
        super().__init__()
        self._service = service
        self._methods = methods
        self._url = url
        self._input_schema = input_schema
        self._output_schema = output_schema
        self._output = output
        self._output_status = output_status

    def _setUp(self):
        if self._output_schema:
            error_list = validate(self._output, self._output_schema)
            if error_list:
                msg = (
                    "While setting up a service mock for the '{s._service}' "
                    "service's '{s._url}' endpoint, the specified output "
                    "does not match the service's endpoint output schema.\n\n"
                    "The errors are:\n{errors}\n\n"
                ).format(s=self, errors='\n'.join(error_list))
                raise AssertionError(msg)

        config = get_service_locations()
        service_location = config.get(self._service)
        if service_location is None:
            raise AssertionError(
                "A service mock for the '%s' service was requested, but the "
                "mock has not been configured with a location for that "
                "service. Ensure set_service_locations has been "
                "called before the mock is required, and that the locations "
                "dictionary contains a key for the '%s' service."
                % (self._service, self._service)
            )

        full_url = urllib.parse.urljoin(service_location, self._url)

        def _callback(request):
            if self._input_schema:
                payload = json.loads(request.body.decode())
                error_list = validate(payload, self._input_schema)
                if error_list:
                    # TODO: raise AssertionError here, since this is in a test?
                    return (
                        400,
                        {'Content-Type': 'application/json'},
                        json.dumps(error_list),
                    )
            # TODO: Do we need to support more than just json responses?
            return (
                self._output_status,
                {"Content-Type": "application/json"},
                json.dumps(self._output)
            )

        if ServiceMock._requests_mock is None:
            ServiceMock._requests_mock = responses.RequestsMock(
                assert_all_requests_are_fired=False
            )
            self.addCleanup(ServiceMock._clean_requests_mock)
            ServiceMock._requests_mock.start()
            self.addCleanup(ServiceMock._requests_mock.stop)

        for method in self._methods:
            self._requests_mock.add_callback(method, full_url, _callback)

    @property
    def calls(self):
        return self._requests_mock.calls

    @classmethod
    def _clean_requests_mock(cls):
        cls._requests_mock = None
