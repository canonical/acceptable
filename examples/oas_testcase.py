"""Test case for OpenAPI specification (OAS) output."""

from acceptable import AcceptableService

service = AcceptableService("OpenApiSample")

foo_api = service.api("/foo/<p:int>/<q>", "foo", introduced_at=2)

foo_api.params_schema = {
    "type": "object",
    "required": ["param1"],
    "properties": {
        "param1": {"type": "string"},
        "param2": {"type": "integer"},
    },
}

foo_api.request_schema = {
    "type": "object",
    "required": ["foo", "baz"],
    "properties": {
        "foo": {"description": "This is a foo.", "type": "string"},
        "baz": {
            "type": "object",
            "description": "Bar the door.",
            "introduced_at": 4,
            "properties": {
                "bar": {"type": "string", "introduced_at": 5, "description": "asdf"}
            },
        },
    },
}

foo_api.response_schema = {
    "type": "object",
    "properties": {
        "foo_result": {"description": "Result of a foo.", "type": "string"},
        "bar": {"type": "string", "description": "bar bar", "introduced_at": 5},
    },
}

foo_api.changelog(5, "Added baz field.")
foo_api.changelog(4, "Added bar field")


@foo_api
def foo():
    """Documentation goes here."""
