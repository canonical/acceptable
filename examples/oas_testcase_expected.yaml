components:
  schemas:
    Default:
      properties:
        code:
          format: int32
          type: integer
      type: object
info:
  contact:
    email: example@example.example
    name: ''
  description: OpenApiSample
  title: OpenApiSample
  version: 0.0.5
openapi: 3.1.0
paths:
  /foo:
    get:
      description: Documentation goes here.
      operationId: foo
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Default'
      responses:
        '200':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Default'
          description: OK
      tags:
      - none
servers:
- url: http://localhost
tags:
- name: none
